### PR TITLE
Problem (Fix #1800): hkdf 0.9 fails to compile

### LINKED ISSUE #1720: Problem: aead 0.3 fails to compile
https://travis-ci.org/github/crypto-com/chain/jobs/694856534#L921

### LINKED ISSUE #1722: Problem: crypto-mac 0.8.0 fails to compile
```
   Compiling ra-enclave v0.5.0 (/drone/src/chain-tx-enclave-next/enclave-ra/ra-enclave)
1273	error[E0277]: the trait bound `cmac::CmacError: std::convert::From<crypto_mac::errors::MacError>` is not satisfied
1274	  --> chain-tx-enclave-next/enclave-ra/ra-enclave/src/cmac.rs:33:9
1275	   |
1276	33 |         inner.verify(&code[..]).map_err(Into::into)
1277	   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<crypto_mac::errors::MacError>` is not implemented for `cmac::CmacError`
1278	   |
1279	   = help: the following implementations were found:
1280	             <cmac::CmacError as std::convert::From<crypto_mac::errors::MacError>>
1281	   = note: required because of the requirements on the impl of `std::convert::Into<cmac::CmacError>` for `crypto_mac::errors::MacError`
1282	   = note: required by `std::convert::Into::into`
1283	
1284	error[E0277]: the trait bound `cmac::CmacError: std::convert::From<crypto_mac::errors::MacError>` is not satisfied
1285	  --> chain-tx-enclave-next/enclave-ra/ra-enclave/src/cmac.rs:33:41
1286	   |
1287	33 |         inner.verify(&code[..]).map_err(Into::into)
1288	   |                                         ^^^^^^^^^^ the trait `std::convert::From<crypto_mac::errors::MacError>` is not implemented for `cmac::CmacError`
1289	   |
1290	   = help: the following implementations were found:
1291	             <cmac::CmacError as std::convert::From<crypto_mac::errors::MacError>>
1292	   = note: required because of the requirements on the impl of `std::convert::Into<cmac::CmacError>` for `crypto_mac::errors::MacError`
1293	   = note: required by `std::convert::Into::into`
1294	
1295	error: aborting due to 2 previous errors
1296	
1297	For more information about this error, try `rustc --explain E0277`.
1298	error: could not compile `ra-enclave`.
```

### LINKED ISSUE #1729: Problem: aes 0.4.0 fails to compile
https://github.com/crypto-com/chain/issues/1720
https://travis-ci.org/github/crypto-com/chain/jobs/695159772#L918

### LINKED ISSUE #1733: Problem: aes-gcm-siv 0.5 fails to compile
https://travis-ci.org/github/crypto-com/chain/jobs/695468690#L984

### LINKED ISSUE #1735: Problem: aes-gcm 0.6 fails to build
```
error[E0599]: no function or associated item named `new` found for struct `aes_gcm::AesGcm<aesni::aes128::Aes128, typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UTerm, typenum::bit::B1>, typenum::bit::B1>, typenum::bit::B0>, typenum::bit::B0>>` in the current scope
1561	  --> chain-tx-enclave-next/enclave-utils/src/lib.rs:74:31
1562	   |
1563	74 |         let aead = Aes128Gcm::new(gk);
1564	   |                               ^^^ function or associated item not found in `aes_gcm::AesGcm<aesni::aes128::Aes128, typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UInt<typenum::uint::UTerm, typenum::bit::B1>, typenum::bit::B1>, typenum::bit::B0>, typenum::bit::B0>>`
1565	   |
1566	   = help: items from traits can only be used if the trait is in scope
...
```

https://github.com/crypto-com/chain/issues/1729

### LINKED ISSUE #1736: Problem: cmac 0.3 fails to compile
https://github.com/crypto-com/chain/issues/1722

### LINKED ISSUE #1749: Problem: digest 0.9 fails to compile
https://travis-ci.org/github/crypto-com/chain/jobs/696684957#L906

bumping together with all other inter-related crates may resolve it: https://github.com/crypto-com/chain/pull/1739

### LINKED ISSUE #1751: Problem: sha2 0.9 fails to compile
https://github.com/crypto-com/chain/pull/1750
https://github.com/crypto-com/chain/issues/1749

### LINKED ISSUE #1757: Problem: ripemd160 0.9 fails to compile
used in HD wallet tests
https://github.com/crypto-com/chain/pull/1755
https://github.com/crypto-com/chain/issues/1749

### LINKED ISSUE #1800: Problem: hkdf 0.9 fails to compile
https://travis-ci.org/github/crypto-com/chain/jobs/700907716#L909
https://github.com/crypto-com/chain/issues/1749

### LINKED ISSUE #1932: Problem: hpke 0.1.7 fails to compile
https://travis-ci.org/github/crypto-com/chain/jobs/707004854#L1142

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- upgrade related dependencies
- Fix #1751
- Fix #1749
- Fix #1757
- Fix #1736
- Fix #1722
- Fix #1720
- Fix #1729
- Fix #1735
- Fix #1733
- Fix #1932

### CODE REVIEW

**Comment by yihuang** [comment]
> not sure why in Cargo.lock there's still "aes 0.3.2" as dependency of ra-enclave?

aes is upgraded now.

**Comment by yihuang** [comment]
> hpke also now feature guards so one can only depend on p256.
> I'm looking at Cargo.lock and some of the old crates are still in there -- some are due to old signatory/tendermint-rs #828
> others due to libra-crypto -- can that be updated?

Upgraded signatory to 0.20 in tendermint-rs, it still uses sha2 0.8:
signatory-dalek 0.20 -> ed25519-dalek -> sha2 0.8

**Comment by tomtau** [comment]
multi-node integration test fails:
```
Error: Run enclave error: enclave runner return error code: None, stderr: 
159	run ra-sp-server beforehand  e.g.) ./ra-sp-server --quote-type Unlinkable --ias-key $IAS_API_KEY --spid $SPID
160	check mls path is correct  e.g.) mls.sgxs, mls.sig  <- two files are necessary
161	Traceback (most recent call last):
162	  File "./multinode/join_test.py", line 73, in <module>
163	    txid = rpc.staking.join(
164	  File "/drone/src/integration-tests/bot/chainrpc.py", line 257, in join
165	    return self.client.call(
166	  File "/drone/src/integration-tests/bot/chainrpc.py", line 99, in call
167	    assert 'result' in rsp, rsp['error']
168	AssertionError: {'code': -32603, 'message': 'Tendermint RPC error: public tx process failed: node join tx process failed: key package decode failed'}
```

**Comment by yihuang** [comment]
> Build failed:
> 
> * [drone-hw/push](http://13.82.172.109/crypto-com/chain/852)

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RaContextError(AesmError(AesmCommunication(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" })))', chain-tx-enclave-next/enclave-ra/ra-sp-server/src/main.rs:15:18
```
ra-sp-server unable to start.

**Comment by tomtau** [comment]
bors retry

**Comment by tomtau** [comment]
bors retry

**Comment by devashishdxt** [comment]
@yihuang There is still a merge conflict in `Cargo.lock`

**Comment by yihuang** [comment]
> @yihuang There is still a merge conflict in `Cargo.lock`

done

**Comment by devashishdxt** [comment]
@yihuang I don't know. But, to me, it still shows that there is a merge conflict.

**Comment by yihuang** [comment]
> @yihuang I don't know. But, to me, it still shows that there is a merge conflict.

done again, it conflict with newer commits.

**Comment by yihuang** [comment]
fixed merge conflict again

**Comment by tomtau** [comment]
@devashishdxt 
the patched ring may need updating?

```
warning: Patch `ring v0.16.13 (https://github.com/crypto-com/ring.git?rev=4e1862fb0df9efaf2d2c1ec8cacb1e53104f3daa#4e1862fb)` was not used in the crate graph.
1243	Check that the patched package version and available features are compatible
1244	with the dependency requirements. If the patch has a different version from
1245	what is locked in the Cargo.lock file, run `cargo update` to use the new
1246	version. This may also occur with an optional dependency that is not enabled.
```

```
error[E0425]: cannot find function `fill_impl` in this scope
1303	   --> /drone/src/drone/cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.15/src/rand.rs:160:9
1304	    |
1305	160 |         fill_impl(dest)
1306	    |         ^^^^^^^^^ help: try: `self.fill_impl`
1307	
1308	   Compiling regex v1.3.9
1309	   Compiling rand_chacha v0.2.2
1310	error: aborting due to previous error
1311	
1312	For more information about this error, try `rustc --explain E0425`.
1313	error: could not compile `ring`.
```

**Comment by devashishdxt** [comment]
> @devashishdxt 
> 
> the patched ring may need updating?
> 
> 
> 
> ```
> 
> warning: Patch `ring v0.16.13 (https://github.com/crypto-com/ring.git?rev=4e1862fb0df9efaf2d2c1ec8cacb1e53104f3daa#4e1862fb)` was not used in the crate graph.
> 
> 1243	Check that the patched package version and available features are compatible
> 
> 1244	with the dependency requirements. If the patch has a different version from
> 
> 1245	what is locked in the Cargo.lock file, run `cargo update` to use the new
> 
> 1246	version. This may also occur with an optional dependency that is not enabled.
> 
> ```
> 
> 
> 
> ```
> 
> error[E0425]: cannot find function `fill_impl` in this scope
> 
> 1303	   --> /drone/src/drone/cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.15/src/rand.rs:160:9
> 
> 1304	    |
> 
> 1305	160 |         fill_impl(dest)
> 
> 1306	    |         ^^^^^^^^^ help: try: `self.fill_impl`
> 
> 1307	
> 
> 1308	   Compiling regex v1.3.9
> 
> 1309	   Compiling rand_chacha v0.2.2
> 
> 1310	error: aborting due to previous error
> 
> 1311	
> 
> 1312	For more information about this error, try `rustc --explain E0425`.
> 
> 1313	error: could not compile `ring`.
> 
> ```

Yes. I’ll update ring.

**Comment by devashishdxt** [comment]
Can you resolve the merge conflics @yihuang?

**Comment by yihuang** [comment]
> Can you resolve the merge conflics @yihuang?

done

**Comment by yihuang** [comment]
> > Can you resolve the merge conflics @yihuang?
> 
> done

integration tests failed locally through.

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RaContextError(AesmError(AesmCommunication(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" })))', chain-tx-enclave-next/enclave-ra/ra-sp-server/src/main.rs:15:18

need longer time?

**Comment by yihuang** [comment]
> > thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RaContextError(AesmError(AesmCommunication(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" })))', chain-tx-enclave-next/enclave-ra/ra-sp-server/src/main.rs:15:18
> 
> need longer time?

This log might not be problematic, it can restart for several times before start successfully, and successful start don't have logs.

File: chain-tx-enclave-next/mls/Cargo.toml [hunk 1]
```diff
@@ -11,11 +11,12 @@ ring = "0.16"
...
 x509-parser = "0.8.0-beta3"
-sha2 = "0.8.1"
-hkdf = "0.8"
+sha2 = "0.9"
+hkdf = "0.9"
 secrecy = "0.6.0"
-hpke = "0.1.5"
-aead = "0.2.0"
+# FIXME change after hpke release
+hpke = { git = "https://github.com/rozbb/rust-hpke.git", branch = "master", default-features = false, features = ["p256"] }
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/Cargo.toml:18]: there's a 0.1.7 release now: https://github.com/crypto-com/chain/issues/1932

Comment (yihuang) [reply chain-tx-enclave-next/mls/Cargo.toml]: done

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
bors try

**Comment by yihuang** [comment]
the integration-test failure is fixed.

**Comment by tomtau** [comment]
bors try

**Comment by yihuang** [comment]
> maybe for now, make sha2 optional in chain-core (with `#[cfg(not(feature = "mesalock_sgx"))]`), as those sha2 operations aren't needed in tx-validation enclave?

yeah, done.

**Comment by tomtau** [comment]
bors try

**Comment by yihuang** [comment]
> ## try
> Build failed:
> 
> * [drone-hw/push](http://13.82.172.109/crypto-com/chain/965)

fixed

**Comment by tomtau** [comment]
bors try

File: chain-core/Cargo.toml [hunk 2]
```diff
@@ -7,16 +7,16 @@ readme = "../README.md"
 edition = "2018"
 
 [features]
-default = ["serde", "bech32", "hex", "base64", "secp256k1/serde", "secp256k1/std", "mls", "ra-client"]
+default = ["sha2", "serde", "bech32", "hex", "base64", "secp256k1zkp/serde", "secp256k1zkp/std", "mls", "ra-client"]
```
Comment (tomtau) [inline chain-core/Cargo.toml:10]:
```suggestion
default = ["sha2", "serde", "bech32", "hex", "base64", "secp256k1/serde", "secp256k1/std", "mls", "ra-client"]
```

Comment (yihuang) [reply chain-core/Cargo.toml]: done

**Comment by tomtau** [comment]
bors try

File: Cargo.toml [hunk 3]
```diff
@@ -52,4 +52,5 @@ default-members = [
 ]
 
 [patch.crates-io]
-ring = { git = "https://github.com/crypto-com/ring.git", rev = "7d08bdbbc68fe5e867445c576c004336699650c1" }
+ring = { git = "https://github.com/crypto-com/ring.git", rev = "bdbcc7041095f028d49d9fecd7edcf26d6083274" }
+sha2 = { git = "https://github.com/yihuang/hashes.git", branch = "master" }
```
Comment (tomtau) [inline Cargo.toml:56]:
```suggestion
# FIXME: use upstream when merged
sha2 = { git = "https://github.com/crypto-com/hashes.git", rev = "289d5b76f2163a3808010341ed1df3cb156d97e1" }
```

Comment (yihuang) [reply Cargo.toml]: done

Comment (yihuang) [reply Cargo.toml]: done

**Pull request merged**